client/src/research/DynamicTable.js

```diff
@@ -0,0 +1,355 @@
+import Immutable from 'immutable';
+import PropTypes from 'prop-types';
+import * as React from 'react';
+
+import moment from 'moment';
+import './ResearcherDataPage.css';
+
+import {Table, Column, SortDirection, SortIndicator, CellMeasurer, CellMeasurerCache} from 'react-virtualized';
+import 'react-virtualized/styles.css';
+import {requestTranscript} from './Transcribe.js';
+
+export default class DynamicHeightTableColumn extends React.PureComponent {
+  static propTypes = {
+    list: PropTypes.instanceOf(Immutable.List).isRequired,
+    width: PropTypes.number.isRequired,
+  };
+
+  constructor(props) {
+    super(props);
+
+    const sortBy = 'index';
+    const sortDirection = SortDirection.ASC;
+    const sortedList = this._sortList({sortBy, sortDirection});
+
+    //Create a dictionary mapping audioID to an audio element with the audio loaded in
```

Comment:
There is a lot happening in the constructor for this component :)

Is there a way to encapsulate the work to load the audio and the transcript in a separate React component?  The point of React is so you don't have to manually manage these kinds of DOM interactions, which get increasingly difficult as you add new features or modify.  In general, if there's a ton going on in the constructor or if you're ever calling `forceUpdate` it usually means you're fighting the framework.

It sounds like one way to factor that would be to have the audio cell in the table use a different render method, that renders `<AudioWithTranscript audioId={audioId} />`.  And then that component could handle fetching the audio and transcript individually for each component as part of its `componentDidMount`.  Then you wouldn't have to manage all these changes to elements and be calling `forceUpdate` and could remove all that code.  And this will make the UX better too - it will incrementally render each component as the transcript loads rather than blocking the whole UI until all the loading is done with the `Promise.all`.

So this would be an awesome thing to do in a refactor phase.

Reply:
Hmm. This makes sense. I know a partially decided to do it this way because I thought it would be annoying to have the table constantly updating with new information as audio and trascripts came in. I figured updating the table twice, once when the audio came in and once when the transcripts came in would be a smoother experience.

I'm trying to remember if there were reasons from the engineering side of things. I'll make an issue so I remember to revisit this!

Reply:
Issue https://github.com/mit-teaching-systems-lab/threeflows/issues/362

Just got through the rest of your comments about this. I definitely need to fix this to use separate react components.